### PR TITLE
fix(compiler): Throw error when variables with identical names are de…

### LINKED ISSUE #6492: declaring two variables within the same name does not fail (should better yell out conflict)
if you define 

``` html
<input #searchTerm/>
<input #searchTerm/>
<button (click)="search(searchTerm)"> Search </button>
```

it does not fail for conflict. 

I think it's better if it does so we can early check a conflict var name.

### PR DESCRIPTION
**Pull request by madskonradsen**
This PR closes #6492 as the compiler now throws an error if two variables with identical names are declared in a template.

### CODE REVIEW

File: modules/angular2/src/compiler/template_parser.ts [hunk 1]
```diff
@@ -486,6 +487,12 @@ class TemplateParseVisitor implements HtmlAstVisitor {
     if (identifier.indexOf('-') > -1) {
       this._reportError(`"-" is not allowed in variable names`, sourceSpan);
     }
+
+    if (this.parsedVariables.indexOf(identifier) > -1) {
+      this._reportError(`Variable "#${identifier}" is defined multiple times`, sourceSpan);
+    }
+    this.parsedVariables.push(identifier);
```
Comment (sandyjoshi) [inline modules/angular2/src/compiler/template_parser.ts]: my view is if variable is found to be duplicate then it should not be pushed in parsedVariables array. 
Let me know if i miss understood anything.

Comment (madskonradsen) [inline modules/angular2/src/compiler/template_parser.ts]: True, that would make sense. On the other hand, both scenarios would throw an error that should be handled by the developer, so why would it make the code better? :)

Comment (sandyjoshi) [inline modules/angular2/src/compiler/template_parser.ts]: Sorry but i don't understood How both scenarios would throw an error ? you will call this._reportError only in one case.

if (this.parsedVariables.indexOf(identifier) > -1) {
  this._reportError(`Variable "#${identifier}" is defined multiple times`, sourceSpan);
} else {
  this.parsedVariables.push(identifier);
  targetVars.push(new VariableAst(identifier, value, sourceSpan));
}

Comment (madskonradsen) [inline modules/angular2/src/compiler/template_parser.ts]: What I'm trying to say; would it make any difference if we have duplicates? The Duplicates will only exist on errors, and in that case, an error will be thrown so the developer can fix it. What benefits would we get by having an else-clause, except from having more code? :)

Comment (sandyjoshi) [inline modules/angular2/src/compiler/template_parser.ts]: ok understood it is not of any use as of now.

File: modules/angular2/test/compiler/template_parser_spec.ts [hunk 2]
```diff
@@ -780,6 +780,12 @@ There is no directive with "exportAs" set to "dirA" ("<div [ERROR ->]#a="dirA"><
 "-" is not allowed in variable names ("<div [ERROR ->]#a-b></div>"): TestComp@0:5`);
         });
 
+        it('should report duplicate variable names', () => {
```
Comment (mhevery) [inline modules/angular2/test/compiler/template_parser_spec.ts]: Please add one more tests, which says that the same variables across different templates are OK.

```
<div #a>
  <span *ngIf="..." #a>OK</span>
</div>
```
The above is allowed.

**Comment by mhevery** [comment]
I am sorry for the delay in getting back to you. 

- [ ] Please add a test as described in my comment
- [ ] Please rebase to the latest master and ensure that all tests pass. (Currently I see two failures which prevent me from merging.)

File: modules/angular2/test/core/linker/query_integration_spec.ts [hunk 3]
```diff
@@ -290,7 +290,7 @@ export function main() {
       it('should contain all content children',
          inject([TestComponentBuilder, AsyncTestCompleter], (tcb: TestComponentBuilder, async) => {
            var template =
-               '<needs-content-children-read #q text="ca"><div #q text="cb"></div></needs-content-children-read>';
```
Comment (tbosch) [inline modules/angular2/test/core/linker/query_integration_spec.ts]: What are the changes in this file about? The should not be required for the change in this PR...

Comment (tbosch) [inline modules/angular2/test/core/linker/query_integration_spec.ts]: Ah, just saw that the variable was used twice.

**Comment by madskonradsen** [comment]
I'll fix it as soon as you guys have the gulp-scripts functioning again(The RC structure-changes broke it)... It's a bit tricky to to test and build stuff without it...

**Comment by mhevery** [comment]
@madskonradsen use `./test.sh` for testing.

**Comment by madskonradsen** [comment]
I can see that you closed the issue yesterday as someone else made a PR. Closing this.

**Pull request closed without merging**